### PR TITLE
Token healing

### LINKED ISSUE #28346: Token healing (under 40 LOC)
### Feature request

Token healing rectifies the token boundary bias in greedy tokenization. It does this by trimming and regrowing the prompt to better align with the model's tokenizer, thus enhancing generation quality. The improvement is clearest with completion models.

Token boundary bias is a silent performance killer that doesn't seem very well known. It has clear impact on completion quality, though I'm not sure where it would fit as a transformers feature.

A more thorough explanation of the problem: [The Art of Prompt Design: Prompt Boundaries and Token Healing | by Scott Lundberg](https://towardsdatascience.com/the-art-of-prompt-design-prompt-boundaries-and-token-healing-3b2448b0be38).

### Motivation

Given a completion prompt with a partial url ending with `:`, the model might have seen the expected completion `://` as a _single_ token in training. However, the prompt's tail token `:` tells it that the next token is not `//`, and so it generates a wrong completion. Such errors compound in auto-regressive language models.

### Your contribution

My implementation (under 40 LOC): https://github.com/Ayenem/TokenHealer/tree/main

### PR DESCRIPTION
**Pull request by ahmed-moubtahij**
# What does this PR do?

Token healing rectifies the token boundary bias in greedy tokenization. It does this by trimming and regrowing the prompt to better align with the model's tokenizer, thus enhancing generation quality. The improvement is clearest with completion models.

Token boundary bias is a silent performance killer that doesn't seem very well known. It has clear impact on completion quality, though I'm not sure where it would fit as a transformers feature.

A more thorough explanation of the problem: [The Art of Prompt Design: Prompt Boundaries and Token Healing | by Scott Lundberg](https://towardsdatascience.com/the-art-of-prompt-design-prompt-boundaries-and-token-healing-3b2448b0be38).

### Motivation

Given a completion prompt with a partial url ending with `:`, the model might have seen the expected completion `://` as a _single_ token in training. However, the prompt's tail token `:` tells it that the next token is not `//`, and so it generates a wrong completion. Such errors compound in auto-regressive language models.

Fixes [#28346](https://github.com/huggingface/transformers/issues/28346)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

- @gante

### CODE REVIEW

**Comment by Ayenem** [comment]
Circleci tests seem to fail, `tests_flax` for example outputs:

> ImportError while importing test module '/home/circleci/transformers/tests/generation/test_token_healing.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> ../.pyenv/versions/3.8.12/lib/python3.8/importlib/__init__.py:127: in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
> tests/generation/test_token_healing.py:23: in <module>
>     class TokenHealingTestCase(unittest.TestCase):
> tests/generation/test_token_healing.py:26: in TokenHealingTestCase
>     completion_model = AutoModelForCausalLM.from_pretrained(
> ../.pyenv/versions/3.8.12/lib/python3.8/site-packages/transformers/utils/import_utils.py:1304: in __getattribute__
>     requires_backends(cls, cls._backends)
> ../.pyenv/versions/3.8.12/lib/python3.8/site-packages/transformers/utils/import_utils.py:1292: in requires_backends
>     raise ImportError("".join(failed))
> E   ImportError: 
> E   AutoModelForCausalLM requires the PyTorch library but it was not found in your environment.

I don't know how to specify to the Circleci environment that it should install pytorch. This is my first PR, so I would appreciate some guidance 🙏

File: examples/pytorch/token-healing/README.md
Comment (gante) [inline examples/pytorch/token-healing/README.md:1]: `examples/research/token-healing` would be a better place to hold this example :) The examples in `examples/pytorch` are selected by the `transformers` team and actively maintained 🤗

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:1171]: This new flag should go [here](https://github.com/huggingface/transformers/blob/de13a951b38b85195984164819f1ab05fe508677/src/transformers/generation/configuration_utils.py#L260) instead, with a corresponding entry in the docstring above.

By placing it in the place I'm suggesting, users can still do `model.generate(..., token_healing=true)`, but it has the following advantages:
1. It's consistent with how we handle options in `generate`
2. Doesn't increase the signature of `generate`
3. Owners of models can save `token_healing=True` in their `generation_config`, so that their users can benefit from it by default if the owner believes it is relevant for the model 🙌

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:1727]: These imports should go to the top. It also means `pygtrie` should be added to  [this list](https://github.com/huggingface/transformers/blob/de13a951b38b85195984164819f1ab05fe508677/setup.py#L97) ⚠️

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:1712]: This one is not needed, as `generate` already has it

```suggestion
```

File: tests/generation/test_token_healing.py
Comment (gante) [inline tests/generation/test_token_healing.py:20]: This import is what's breaking CI: the test fetcher for all CI runs has to execute this line and, for test images without pytorch installed, this line will fail :)

Read my comment below for a suggested fix

File: tests/generation/test_token_healing.py
Comment (gante) [inline tests/generation/test_token_healing.py:23]: I'm going to ask to move this test class to `tests/generation/test_utils.py`. By default, we only create a new file when there is significant content for the file in question 🤗 

The imports on the file I'm suggesting are correct, so it remove the CI issues!

**Review by gante — COMMENTED** [review]
Thank you for opening this cool PR! I have written a few comments, but they should be easy to solve 🤗

**Comment by Ayenem** [comment]
> Thank you for opening this cool PR! I have written a few comments, but they should be easy to solve 🤗

Thank you for your detailed reviews and for taking the time to explain their reasons 🙏 Hopefully, they're all fixed now.

File: src/transformers/generation/utils.py
Comment (gante) [inline src/transformers/generation/utils.py:37]: This one should be `from ..models.auto.tokenization_auto import AutoTokenizer` instead

**Review by gante — APPROVED** [review]
Thank you for iterating 🤗 

To fix the dependency issue, I believe you need to add `pygtrie` to `extras["torch"]` in `setup.py` (L262). Otherwise, `pygtrie` is not installed -- being in the list where you added the dependency is not enough to install the dependency :)

**Comment by Ayenem** [comment]
> Thank you for iterating 🤗
> 
> To fix the dependency issue, I believe you need to add `pygtrie` to `extras["torch"]` in `setup.py` (L262). Otherwise, `pygtrie` is not installed -- being in the list where you added the dependency is not enough to install the dependency :)

Thank you for the follow-up! I applied your suggestions. CI is still breaking on the pygtrie import unfortunately. I'll try to think about it some more this weekend.

File: setup.py
Comment (gante) [inline setup.py:264]:
```suggestion
extras["torch"] = deps_list("torch", "accelerate", "pygtrie")
```

Try this instead :) The `extras["pygtrie"]` you defined is not installed, as it is not included in `extras["all"]`

File: tests/generation/test_utils.py
Comment (gante) [inline tests/generation/test_utils.py:3471]:
```suggestion
```

File: tests/generation/test_utils.py
Comment (gante) [inline tests/generation/test_utils.py:3488]:
```suggestion
    def test_prompts(self, name, input, expected):
        model_name_or_path = "TheBloke/deepseek-llm-7B-base-GPTQ"
        tokenizer = AutoTokenizer.from_pretrained(model_name_or_path, use_fast=True)
        completion_model = AutoModelForCausalLM.from_pretrained(
            model_name_or_path,
            device_map="auto",
            trust_remote_code=False,
            revision="main",
            use_cache=True,
        )
```

**Comment by gante** [comment]
@Ayenem hope you don't mind, moving a few minor things around to make our CI happy 🤗

**Comment by Ayenem** [comment]
> @Ayenem hope you don't mind, moving a few minor things around to make our CI happy 🤗

Thank you for taking the time!

File: examples/flax/_tests_requirements.txt
Comment (gante) [inline examples/flax/_tests_requirements.txt:9]:
```suggestion
torch
accelerate
```

**Comment by gante** [comment]
`examples_flax` is failing for reasons unrelated to this PR, I'm going to see if I can open a separate PR to fix it :)

**Comment by gante** [comment]
Ok, actually, it has to be fixed in this PR: `pygtrie` has to be added to `/examples/flax/_tests_requirements.txt` :( 

As for the other failures, I recommend the following:
1. Rebase with `main`
2. run `make fixup` in the `transformers` folder
3. commit the resulting changes

If issues persist, I'll dive into the failures :)

**Comment by gante** [comment]
uhmmm something has gone wrong, the diff shouldn't be 18k lines long 😅

**Comment by Ayenem** [comment]
> uhmmm something has gone wrong, the diff shouldn't be 18k lines long 😅

Indeed...I did a `git fetch upstream` followed by `git rebase upstream/main` then solved conflicts and pushed. I suppose I misinterpreted what you meant by rebasing with main 😓

**Comment by gante** [comment]
@Ayenem That should be it: 
```
git remote add upstream https://github.com/huggingface/transformers.git
git fetch upstream
git rebase upstream/main
(handle merge conflicts)
git push origin your_branch -f
```

possibly something went wrong in the merge conflicts stage?

**Comment by gante** [comment]
@Ayenem you're getting the best of CI problems, hehe

**Comment by Ayenem** [comment]
> @Ayenem you're getting the best of CI problems, hehe

Lucky me 😄 Rebase should've worked as expected now 🤞

**Comment by gante** [comment]
@Ayenem looking at the test errors, other than sorting the merge conflicts (which should fix most of the problems), the new test needs `@require_auto_gptq`. Alternatively, a smaller model like `gpt2` can be used.

**Comment by Ayenem** [comment]
> @Ayenem looking at the test errors, other than sorting the merge conflicts (which should fix most of the problems), the new test needs `@require_auto_gptq`. Alternatively, a smaller model like `gpt2` can be used.

I don't seem to have access to those conflicts locally and I don't have write access for resolving them from github 😕 
image

**Comment by gante** [comment]
@Ayenem I have no more suggestions as well. When this happens, it often means something unexpected happened at rebase time. Most (or even all) CI errors we're seeing here are also already solved on `main` 😞 

May I suggest to:
1. Close this PR
2. Update `main` on your fork
3. Create a new branch on your fork, past these changes, and open a new PR?

Again, apologies for all this additional work 🤗

**Pull request closed without merging**
